File: Leetcode/1996/1996.cpp
bool cmp(const pair<int,int>& l, const pair<int,int>&r) {
    return l.first < r.first;
}

class Solution {
public:
    // O(nlogn)
    int numberOfWeakCharacters(vector<vector<int>> &properties) {
        int n = (int)properties.size();
        vector<pair<int,int>> A(n);
        for (int i = 0; i < n; ++i) {
            A[i].first = properties[i][0];
            A[i].second = properties[i][1];
        }
        sort(A.begin(),A.end(),cmp);
        vector<int> maxFromRight(n);
        maxFromRight[n-1] = A[n-1].second;
        for (int i = n-2; i >= 0; --i) {
            maxFromRight[i] = max(maxFromRight[i+1], A[i].second);
        }
        int counter = 0;
        pair<int,int> temp;
        temp.second = -1;
        for (int i = 0; i < n; ++i) {
            temp.first = A[i].first;
            auto iter = upper_bound(A.begin(),A.end(),temp,cmp);
            if (iter != A.end() && A[i].second < maxFromRight[(int)(iter-A.begin())]) {
                ++counter;
            }
        }
        return counter;
    }
};

File: Leetcode/1996/1996.go
type Pair struct {
	a, b int
}
func numberOfWeakCharacters(properties [][]int) int {
	A := make([]Pair, len(properties))
	for i := 0; i < len(properties); i++ {
		A[i].a = properties[i][0]
		A[i].b = properties[i][1]
	}
	sort.Slice(A, func(i, j int) bool {
		return A[i].a < A[j].a
	})
	maxFromRight := make([]int, len(A))
	maxFromRight[len(A)-1] = A[len(A)-1].b
	for i := len(A)-2; i >= 0; i-- {
		if maxFromRight[i+1] > A[i].b {
			maxFromRight[i] = maxFromRight[i+1]
		} else {
			maxFromRight[i] = A[i].b
		}
	}
	counter := 0
	temp := Pair{0, -1}
	for i := 0; i < len(A); i++ {
		temp.a = A[i].a
		pos := sort.Search(len(A), func(i int) bool {
			return A[i].a > temp.a
		})
		if pos != len(A) && A[i].b < maxFromRight[pos] {
			counter++
		}
	}
	return counter
}